File: runtime/cpp/tests/serials.cpp
#include <nightseam/runtime/peer.hpp>
#include <fstream>
#include <future>
#include <iostream>
#include <thread>
using namespace nightseam::runtime;
using namespace nightseam::duplex;
using namespace std::chrono_literals;
static void require(bool value,const char* message) { if(!value) throw std::runtime_error(message); }
int main(int argc,char** argv) {
    try {
        require(argc==2,"table directory required");
        std::ifstream input(std::string(argv[1])+"/serials.json");
        std::string json((std::istreambuf_iterator<char>(input)),{});
        auto table=parse_value(json);
        for(const auto& row:table.at("rows").array_range()) {
            auto [raw,carrier]=pipe();
            PeerOptions options; options.handlers["echo"]=[](const RequestContext&,Peer&,const Value& value){return value;};
            Peer peer(carrier,Role::server,options);
            auto before=row.at("before").as<std::string>();
            raw->send({Kind::text,before});
            if(parse_value(before).at("kind").as<std::string>()=="request") raw->receive(Wait::after(1s));
            raw->send({Kind::text,row.at("frame").as<std::string>()});
            if(row.at("valid").as<bool>()) {
                raw->receive(Wait::after(1s)); require(!peer.ended(),"valid serial closed peer");
            } else require(peer.await_close(Wait::after(1s)).code==4011,"non-increasing serial admitted");
            peer.close();
        }
        {
            auto [raw,carrier]=pipe();
            std::promise<void> reserved,release;
            auto released=release.get_future().share();
            PeerOptions options; options.max_pending_requests=1;
            options.observer=[&](const Observation& event) {
                if(event.type=="request.started" && !event.incoming) { reserved.set_value(); released.wait(); }
            };
            Peer peer(carrier,Role::client,options);
            auto first=std::async(std::launch::async,[&]{try {peer.call("first");} catch(const std::exception&) {}});
            reserved.get_future().wait();
            bool busy=false;
            try {peer.call("over-budget");} catch(const PublicError& error) {busy=error.code=="busy";}
            release.set_value(); peer.close(); first.get();
            require(busy,"publication waiters escaped the pending budget");
        }
        auto [raw,carrier]=pipe();
        std::promise<void> reserved,release;
        auto released=release.get_future().share();
        PeerOptions options; options.queue_capacity=1;
        options.observer=[&](const Observation& event) {
            if(event.type=="request.started" && !event.incoming && event.id=="c:1") { reserved.set_value(); released.wait(); }
        };
        Peer peer(carrier,Role::client,options);
        auto call=[&]{ try {peer.call("probe");} catch(const std::exception&) {} };
        auto first=std::async(std::launch::async,call);
        reserved.get_future().wait();
        auto second=std::async(std::launch::async,call);
        std::this_thread::sleep_for(50ms); release.set_value();
        auto one=decode_envelope(raw->receive(Wait::after(1s)).data,Role::server);
        auto two=decode_envelope(raw->receive(Wait::after(1s)).data,Role::server);
        peer.close(); first.get(); second.get();
        require(one.id=="c:1" && two.id=="c:2","request publication inverted serials");
        std::cout << "request serial invariants passed\n";
    } catch(const std::exception& e) { std::cerr << e.what() << '\n'; return 1; }
}

File: runtime/rs/tests/serials.rs
use bitwire::Payload;
use nightseam::{Options, Peer, Role};
use nightseam_duplex::{Frame, pipe};
use serde_json::Value;
use std::time::Duration;
use tokio::time::timeout;

#[tokio::test]
async fn shared_serial_table_includes_completed_requests() {
    let table: Value =
        serde_json::from_str(include_str!("../../../conformance/tables/serials.json")).unwrap();
    for row in table["rows"].as_array().unwrap() {
        let (raw, carrier) = pipe(0);
        let peer = Peer::over(carrier, Role::Server, Options::default()).unwrap();
        peer.handle("echo", |_, value| async { Ok(value) }).unwrap();
        let before = row["before"].as_str().unwrap();
        raw.send(Frame::Text(before.as_bytes().to_vec()))
            .await
            .unwrap();
        if serde_json::from_str::<Value>(before).unwrap()["kind"] == "request" {
            timeout(Duration::from_secs(1), raw.receive())
                .await
                .unwrap()
                .unwrap();
        }
        raw.send(Frame::Text(
            row["frame"].as_str().unwrap().as_bytes().to_vec(),
        ))
        .await
        .unwrap();
        if row["valid"] == true {
            let frame = timeout(Duration::from_secs(1), raw.receive())
                .await
                .unwrap()
                .unwrap();
            assert!(matches!(frame, Frame::Text(_)), "{}", row["name"]);
        } else {
            assert_eq!(
                timeout(Duration::from_secs(1), peer.wait_closed())
                    .await
                    .unwrap()
                    .code,
                4011,
                "{}",
                row["name"]
            );
        }
        peer.close();
    }
}

#[tokio::test(flavor = "multi_thread", worker_threads = 4)]
async fn concurrent_calls_publish_increasing_serials() {
    let (raw, carrier) = pipe(0);
    let peer = Peer::over(
        carrier,
        Role::Server,
        Options {
            queue_capacity: 1,
            ..Options::default()
        },
    )
    .unwrap();
    let calls: Vec<_> = (0..64)
        .map(|_| peer.call("probe", Payload::Absent))
        .collect();
    let mut previous = 0;
    for _ in 0..calls.len() {
        let Frame::Text(bytes) = timeout(Duration::from_secs(1), raw.receive())
            .await
            .unwrap()
            .unwrap()
        else {
            panic!("text request expected")
        };
        let frame: Value = serde_json::from_slice(&bytes).unwrap();
        let serial: u64 = frame["id"].as_str().unwrap()[2..].parse().unwrap();
        assert!(serial > previous, "published {serial} after {previous}");
        previous = serial;
    }
    peer.close();
}
